std/string.cup

```diff
@@ -0,0 +1,268 @@
+import "std/vector.cup"
+
+// If you need to debug stuff enable this
+let __SLICE_LIFETIME_PROTECTION: int = false;
+
+struct str{
+    chars: char*;
+    len: i32;
+    isdropped: i32;
+}
+
+// Has the same lifetime as the string it was created from
+// If you drop a string and still use slices you will probably get undefined behaviour
+struct slice{
+    start: char*;
+    len: i32;
+    string_ptr: str*;
+}
+
+// Only to be used internally
+initialize slice(start: char*, len: i32, str_ptr: str*){
+    self.start = start;
+    self.len = len;
+    self.string_ptr = str_ptr;
+}
+
+// Partial lifetime protection
+method slice::is_str_alive(){
+    let str_segment: __MEM_SEGMENT* = __find_segment_by_ptr(self.string_ptr);
+    let is_out_living = false;
+    if (str_segment == null){
+        is_out_living = true;
+    } else {
+        if (!str_segment.istaken){
+            is_out_living = true;
+        }
+    }
+    
+    if (!is_out_living){
+        if (self.string_ptr.isdropped){
+            is_out_living = true;
+        }
+    }
+    
+    if (is_out_living)
+        die(here, "slice: slice outlives string it was made from");
+}
+
+method slice::get(idx: i32): char{
+    if (__SLICE_LIFETIME_PROTECTION)
+        self::is_str_alive();
+    
+    if (idx < self.len){
+        return self.start[idx];
+    } else {
+        die(here, "slice::get: index out of bounds");
+    }
+}
+
+initialize str(chars: char*){
+    self.chars = malloc(sizeof(char)*strlen(chars)+1);
+    self.len = strlen(chars);
+    strcpy(self.chars, chars);
+}
+
+// Creates a str out of slice
+method slice::to_str(): str*{
+    if (__SLICE_LIFETIME_PROTECTION)
+        self::is_str_alive();
+    
+    let chars: char* = malloc(sizeof(char)*self.len+1);
+    
+    for (let i = 0; i < self.len; ++i){
+        chars[i] = self::get(i);
+    }
+    
+    chars[self.len] = '\0';
+    let str_created = new str(chars);
+    free(chars);
+    return str_created;
+}
+
+method str::push_cstr(chars: char*){
+    self.chars = realloc(self.chars,(sizeof(char)*self.len)+(sizeof(char)*strlen(chars))+1);
+    self.len = self.len + strlen(chars);
+    strcat(self.chars, chars);
+}
+
+method str::push(string: str*){
+    self::push_cstr(string.chars);
+}
+
+method str::push_slice(sl: slice*){
+    self.chars = realloc(self.chars,(sizeof(char)*self.len)+(sizeof(char)*sl.len)+1);
+    let first = self.len;
+    self.len = self.len + sl.len;
+    
+    for (let i = first; i < self.len; ++i){
+        self.chars[i] = sl::get(i-first);
+    }
+    
+    self.chars[self.len] = '\0';
+}
+
+method str::len(): i32{
+    return self.len;
+}
```

Comment:
Is there any point to having this helper as opposed to just doing `str.len` where you want it?

Reply:
It looks nicer.